test: enabling wallet migration functional test on windows

m3dwards opened the pull request:
This PR updates the `test/get_previous_releases.py` script to also work on Windows by changing to be pure python rather than using unix tools such as `curl` and `tar`.

This enables additional functional tests to run such as `wallet_migration.py`, `mempool_compatability.py` and `wallet_backwards_compatibility.py`.

Unfortunately `feature_unsupported_utxo_db.py` _could_ run but this test requires Bitcoin `v0.14.3` which will not run under windows with emojis in the data directory (as the functional test runner has by default) . This test could be run as it's own step in the ci workflow file and would pass but as it's quite an old version / feature I have assumed it's not worth worrying about and best just to exclude.

Two tests needed to be slightly modified to run under windows. Both were issues with trying to overwrite a file that already exists which windows seems to be more strict on than the unix based systems.

Finally, building from source has been dropped from the `get_previous_releases.py` script. This had not been updated after the move to cmake and so it was assumed that nobody could have been using that feature.

hebasto reviewed (commented):
Concept ACK.

m3dwards commented on test/get_previous_releases.py:
```diff
@@ -119,55 +135,92 @@ def pushd(new_dir) -> None:
-    tarball = 'bitcoin-{tag}-{platform}.tar.gz'.format(
-        tag=tag[1:], platform=platform)
-    tarballUrl = 'https://bitcoincore.org/{bin_path}/{tarball}'.format(
-        bin_path=bin_path, tarball=tarball)
 
-    print('Fetching: {tarballUrl}'.format(tarballUrl=tarballUrl))
+    host = args.platform
+    if tag < "v23" and host in ["x86_64-apple-darwin", "arm64-apple-darwin"]:
+        host = "osx64"
 
-    ret = subprocess.run(['curl', '--fail', '--remote-name', tarballUrl]).returncode
-    if ret:
-        return ret
+    archive_format = 'tar.gz'
+    if host == 'win64':
+        archive_format = 'zip'
+
+    archive = 'bitcoin-{tag}-{platform}.{archive_format}'.format(
+        tag=tag[1:], platform=host, archive_format=archive_format)
+    archiveUrl = 'https://bitcoincore.org/{bin_path}/{archive}'.format(
+        bin_path=bin_path, archive=archive)
+
+    print('Fetching: {archiveUrl}'.format(archiveUrl=archiveUrl))
+    try:
+        download_from_url(archiveUrl, archive)
+    except Exception as e:
+        print(f"\nDownload failed: {e}")
+        return 1
 
     hasher = hashlib.sha256()
-    with open(tarball, "rb") as afile:
+    with open(archive, "rb") as afile:
         hasher.update(afile.read())
-    tarballHash = hasher.hexdigest()
+    archiveHash = hasher.hexdigest()
 
-    if tarballHash not in SHA256_SUMS or SHA256_SUMS[tarballHash]['tarball'] != tarball:
-        if tarball in [v['tarball'] for v in SHA256_SUMS.values()]:
+    if archiveHash not in SHA256_SUMS or SHA256_SUMS[archiveHash]['archive'] != archive:
+        if archive in [v['archive'] for v in SHA256_SUMS.values()]:
```
Thanks for the tip. As it now deals with zip files too it felt wrong to call them all tarballs.

m3dwards commented:
> For reference, you can also test locally (or in a VM, or in the cloud)

This will probably be better in this instance as it wasn't a good test anyway as the get_previous_releases.py script just used cached versions of the downloads rather than downloading again.

hodlinator reviewed (commented):
Concept ACK 12248ee08026263b70b2a0e6857552c475614207

m3dwards commented on test/get_previous_releases.py:
```diff
@@ -172,73 +226,30 @@ def download_binary(tag, args) -> int:
     return 0
 
 
-def build_release(tag, args) -> int:
-    githubUrl = "https://github.com/bitcoin/bitcoin"
-    if args.remove_dir:
-        if Path(tag).is_dir():
-            shutil.rmtree(tag)
-    if not Path(tag).is_dir():
-        # fetch new tags
-        subprocess.run(
-            ["git", "fetch", githubUrl, "--tags"])
-        output = subprocess.check_output(['git', 'tag', '-l', tag])
-        if not output:
```
@achow101 @maflcko do you or do you know of any use of using the `get_previous_releases.py` script to build from source? Am I being a bit eager in dropping this feature?

m3dwards replied:
@hodlinator thanks for the confirmation.

The other factor worth considering is that as written, this function will not work on Windows. I'm inclined to still leave it out.

m3dwards replied:
@hodlinator agreed and taken

m3dwards commented on test/get_previous_releases.py:
```diff
@@ -92,55 +104,97 @@ def pushd(new_dir) -> None:
+                              '--strip-components=1',
+                              'bitcoin-{tag}'.format(tag=tag[1:])]).returncode
+        if ret != 0:
+            print(f"Failed to extract the {tag} tarball")
```
Taken and applied everywhere.

hodlinator reviewed (commented):
Code review 691e5ba234936020ee47c8f467645bf245ab9fc7

Thanks for incorporating my feedback so far!

Did some testing on Windows. Main concern is about introducing code using legacy python interface.

hodlinator commented on test/get_previous_releases.py:
```diff
@@ -92,55 +104,96 @@ def pushd(new_dir) -> None:
+        percent = min(100, (count * block_size * 100) / total_size)
+        bar_length = 40
+        filled_length = int(bar_length * percent / 100)
+        bar = '█' * filled_length + '-' * (bar_length - filled_length)
```
`$ PYTHONUTF8=1 py test/get_previous_releases.py` worked but I would lean slightly towards ASCII as the error message was so bad (could be mistaken for the downloaded buffer itself somehow being encoded wrongly).

fanquake commented on test/functional/wallet_migration.py:
```diff
@@ -485,11 +485,16 @@ def test_encrypted(self):
         assert_raises_rpc_error(-4, "Error: Wallet decryption failed, the wallet passphrase was not provided or was incorrect", self.migrate_and_get_rpc, "encrypted")
+
         # Use the RPC directly on the master node for the rest of these checks
+        self.master_node.bumpmocktime(1) # Prevents filename duplication on wallet backups as they contain the time in the filename
```
In c78f183800db663bba5f9d3a7280583127eebe3f: It would be good for the commit messages and/or the added doc, to explain that this is (I'm assuming?) only an issue on Windows. If it isn't, and it effects all other platforms, then these test changes could be split out, with a reproducer showing the issue.

m3dwards replied:
It was only a problem on Windows. Made a note of that.

m3dwards commented on test/get_previous_releases.py:
```diff
@@ -92,55 +104,96 @@ def pushd(new_dir) -> None:
             print('Using cached {}'.format(tag))
             return 0
         shutil.rmtree(tag)
     Path(tag).mkdir()
```
Cleaner, taken, thanks!

m3dwards commented on test/get_previous_releases.py:
```diff
@@ -92,55 +104,96 @@ def pushd(new_dir) -> None:
+            sys.stdout.write(f'\rDownloading: [{bar}] {percent:.1f}%')
+            sys.stdout.flush()
+            last_print_time = now
+    urllib.request.urlretrieve(url, archive, reporthook=progress_hook)
```
Good spot, shame it's more verbose but it's all easy code to understand so taken. Thanks for not only noticing but also giving an alternative.

hodlinator replied:
Thanks for taking my suggestion!

m3dwards commented on test/get_previous_releases.py:
```diff
@@ -158,87 +211,44 @@ def download_binary(tag, args) -> int:
-    return 0
-
-
 def check_host(args) -> int:
```
Done.

hodlinator reviewed (approved):
ACK f4e259b31f5785190c4912975173c5025dfe349c

### Concept

Enables more functional tests along with script for fetching previous release binaries on Windows. Removes script support for building previous releases from source on all platforms, which is arguably tricky as it requires having the correct (versions of) dependencies available, and seems to not be well used.

### Changes since previous review

Incorporates feedback from https://github.com/bitcoin/bitcoin/pull/32219#pullrequestreview-2910271020 and https://github.com/bitcoin/bitcoin/pull/32219#discussion_r2137463742.

maflcko reviewed (commented):
left some nits only

review ACK f4e259b31f5785190c4912975173c5025dfe349c 🚿

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK f4e259b31f5785190c4912975173c5025dfe349c 🚿
AHFm4z0J8FOrNMufHlWQ+gS455X6PI9FAbG/jgewgMnYxXsvBevGetO359R3f3LD+FWk20z1Z/fB6VSgKGJGDw==
```

</details>

m3dwards committed:
test: remove building from source from get prev releases script

Using the get_previous_releases.py script to build from source only works for
releases prior to v29 due to removal of Autotools (in favor of CMake). It also
does not support building on Windows, and we are adding support for downloading
Windows release binaries in later commits of this PR.

As there were no complaints during review, it is assumed nobody uses this
functionality.

m3dwards committed:
scripted-diff: rename tarball to archive

Windows zip files are added in the next commit which are not tarballs so
renaming tarball to the more generic term archive which can cover both.

-BEGIN VERIFY SCRIPT-
sed -i 's/tarball/archive/g' test/get_previous_releases.py
-END VERIFY SCRIPT-

m3dwards committed:
test: add windows support to get previous releases script

Needed to allow the wallet migration functional test to run under native
windows. Platform variable renamed to not clash with platform package.

m3dwards committed:
test: remove mempool.dat before copying

Renaming mempool.dat from old node to new node fails on windows as the
file already exists.

m3dwards committed:
ci: disable feature_unsupported_utxo_db functional test

This test required v0.14.3 which doesn't support special characters like
emojis in the datapath on windows. Functional test runner includes a few
emojis in it's default data directory.

m3dwards committed:
test: increment mocked time for migrating wallet backups

The wallet backups performed before migration use the time as part of
their filename. As the time is mocked, increment it between migration
attempts to prevent file name conflicts which is a problem on Windows.

m3dwards committed:
ci: run get_previous_releases as part of test cross win job

This should enable wallet_migration functional test to run on Windows

m3dwards commented on ci/test/03_test_script.sh:
```diff
@@ -99,7 +99,7 @@ if [ -z "$NO_DEPENDS" ]; then
 fi
 if [ "$DOWNLOAD_PREVIOUS_RELEASES" = "true" ]; then
-  test/get_previous_releases.py -b -t "$PREVIOUS_RELEASES_DIR"
+  test/get_previous_releases.py -t "$PREVIOUS_RELEASES_DIR"
```
It was `--target-dir`. Applied suggestion.

m3dwards commented on test/get_previous_releases.py:
```diff
@@ -92,31 +104,73 @@ def pushd(new_dir) -> None:
+    print(f'Fetching: {archiveUrl}')
+
+    try:
+        download_from_url(archiveUrl, archive)
```
Done.

m3dwards commented on test/get_previous_releases.py:
```diff
@@ -92,31 +104,73 @@ def pushd(new_dir) -> None:
+                progress_hook(progress_bytes, total_size)
+
+    print('\n', end="")
+    sys.stdout.flush() # Flush to avoid error output on the same line.
```
Sure, done.

m3dwards commented on test/get_previous_releases.py:
```diff
@@ -92,31 +104,73 @@ def pushd(new_dir) -> None:
+        bar = '#' * filled_length + '-' * (bar_length - filled_length)
+        if now - last_print_time >= 1 or percent >= 100:
+            sys.stdout.write(f'\rDownloading: [{bar}] {percent:.1f}%')
+            sys.stdout.flush()
```
Done.

m3dwards commented on .github/workflows/ci.yml:
```diff
@@ -372,6 +372,14 @@ jobs:
+
+      - name: Get previous releases
+        working-directory: test
+        run: ./get_previous_releases.py -t $PREVIOUS_RELEASES_DIR
```
Changed to `--target-dir`.

hodlinator reviewed (approved):
re-ACK 941b8f54c0d35d3243bb6083f3b52681d1b9a555

Fixed some nits (https://github.com/bitcoin/bitcoin/pull/32219#pullrequestreview-2959123050) since my first ACK (https://github.com/bitcoin/bitcoin/pull/32219#pullrequestreview-2921071060).
Passed linter locally.

m3dwards commented:
@maflcko addressed nits and force pushed

achow101 commented:
ACK 8cc9845b8ddf4f93a02c622e7df8d1095dc1a640

achow101 commented:
ACK 941b8f54c0d35d3243bb6083f3b52681d1b9a555

Sjors commented on test/get_previous_releases.py:
```diff
@@ -107,40 +107,40 @@ def download_binary(tag, args) -> int:
-    Path(tarball).unlink()
+    Path(archive).unlink()
 
     if tag >= "v23" and platform == "arm64-apple-darwin":
```
1a1b478ca31be1f670754a47da17863271e46b7b: missed one occurance of `platform`, see #32922
